### PR TITLE
Add a Returned state for release requests

### LINKED ISSUE #326: Add RETURNED state
To improve the workflow around the back-and-forth between researchers and output checkers, we want to restrict what each type of user can do at certain times.  This means formalising the concept of a "round of review", and adding a new state for a review request: `RETURNED`.

While a request is in this state, output checkers should not be able to interact with a request.  And researchers should only be able to update a request while the the request is in the `PENDING` or `RETURNED` state.

Moving a request to the `RETURNED` state should require a deliberate action by an output checker.  This should only be available once every file in the request has approval decisions from two output checkers in any round, or any two decisions in the current round.  Although this information could be reconstructed from the audit log, it may be better to have a way of explicitly identifying decisions made in the current round.

A notification should be sent when a request moves to the `RETURNED` state; this should send an email to the researcher and update the corresponding GitHub issue. (This will also require some changes to job-server)

A researcher should be able to withdraw a request at any point before it is released.  There does not appear to be possible via the UI, but the `RequestStatus` state machine should be updated to allow withdrawing a request in the `RETURNED` state.

Addressing this will require updating the documentation (perhaps to include a state diagram).  Changes to the documentation should be reviewed by a member of the output checking team.

### PR DESCRIPTION
**Pull request by rebkwok**
Fixes #326. Currently there's no way to return a request from the UI, buttons to do that are left for a later PR (#388).

After a request is submitted, at least 2 reviewers must review each file. At that point, it can be approved/released if all the reviews are approvals. Otherwise, it can be returned. (Note: currently a request will all approved files can also be returned - tbc if we want to allow that.)
Once returned, the request moves into RETURNED state. The researcher can update files (or not) and re-submit it.  This allows for a researcher to re-submit without making any changes. The assumption is that they could have answered questions to provide clarification, but not feel that the file required updating. 

Any updated files are actually new RequestFile objects, so have no reviews and automatically require review when the request is re-submitted. 

Any unchanged files that are approved do not need re-review, and their review status remains unchanged. Any unchanged files that were rejected have their status set to UNDECIDED for that reviewer. This is done by Airlock at the point that the request is re-submitted.

### CODE REVIEW

**Comment by rebkwok** [comment]
We now have methods to reset a file review (i.e. reviewer changes their mind after either approving or rejecting) and to mark a file as undecided (done by Airlock when a request is resubmitted, for files that were rejected in the previous round). I'm not sure "undecided" is a good name - it seemed fine at the time, but I wrote this PR before the reset-review was done, but I'm not sure what's a better one.

**Review by bloodearnest — APPROVED** [review]
Looks good, thanks

I'm not 100% sure about the UNDECIDED flow, it feels a bit awkward.

I wonder if adding a slightly different new state to indicate neither APPROVED or REJECTED might end up cleaner? e.g. CLARIFY, where a checker leaves a comment, but doesn't outright APPROVE or REJECT. We've talked about adding this middle ground vote before.

It counts as a vote in terms of being able to RETURN, but going back to SUBMITTED doesn't require any changes (maybe requires a comment on the group, but that's a bit awkward to enforce).

Than means the REJECTED vote is used explicitly to mean this file is disclosive and *cannot* be released, and thus would block re-SUBMITTING until it has been updated?

Hmm, I wonder if DISCLOSIVE is a better state name than REJECTED?

That said, I am happy to land as is, we can always refactor.

File: tests/unit/test_business_logic.py [hunk 1]
```diff
@@ -579,8 +579,10 @@ def test_provider_get_current_request_for_user_output_checker(bll):
         (RequestStatus.SUBMITTED, RequestStatus.APPROVED, False, True),
         (RequestStatus.SUBMITTED, RequestStatus.REJECTED, False, True),
         (RequestStatus.SUBMITTED, RequestStatus.WITHDRAWN, True, False),
-        (RequestStatus.SUBMITTED, RequestStatus.PENDING, True, False),
```
[the remark this answers is not in the document]
Comment (rebkwok) [reply tests/unit/test_business_logic.py:582]: yes, will add

**Comment by rebkwok** [comment]
I did initially consider a different state for approving, but (a) Peter said that was not part of this ticket and (b) it means that when you re-submit, there's no way to tell that a reviewer has made a new decision. I.e. if they mark a file as "CLARIFY" (or whatever), and it stays as "CLARIFY" when resubmitted, and "CLARIFY" is a good enough decision for returning, there's no way for a reviewer to say that they still need clarification. I think it would still require the approval status to be reset in some way on re-submitting so that the next round requires an explicit decision on all non-approved files.

**Comment by bloodearnest** [comment]
> I did initially consider a different state for approving, but (a) Peter said that was not part of this ticket and (b) it means that when you re-submit, there's no way to tell that a reviewer has made a new decision. I.e. if they mark a file as "CLARIFY" (or whatever), and it stays as "CLARIFY" when resubmitted, and "CLARIFY" is a good enough decision for returning, there's no way for a reviewer to say that they still need clarification. I think it would still require the approval status to be reset in some way on re-submitting so that the next round requires an explicit decision on all non-approved files.

Good point, we will a status reset of some sort either way, and deferring until we implement the new vote state makes sense to me.

**Pull request merged**